src/task/task_environment.rs

```diff
@@ -116,28 +115,21 @@ impl<'p, D: TaskDisambiguation<'p>> SearchEnvironments<'p, D> {
         name: &str,
         source: FindTaskSource<'p>,
     ) -> Result<TaskAndEnvironment<'p>, FindTaskError> {
-        // If the task was specified on the command line and there is no explicit environment and
-        // the task is only defined in the default feature, use the default environment.
+        // If no explicit environment was specified
         if matches!(source, FindTaskSource::CmdArgs) && self.explicit_environment.is_none() {
-            if let Some(task) = self
-                .project
-                .manifest
-                .default_feature()
-                .targets
-                .resolve(self.platform)
-                .find_map(|target| target.tasks.get(name))
-            {
-                // None of the other environments can have this task. Otherwise, its still
-                // ambiguous.
-                if !self
-                    .project
-                    .environments()
-                    .into_iter()
-                    .flat_map(|env| env.features(false).collect_vec())
-                    .flat_map(|feature| feature.targets.resolve(self.platform))
-                    .any(|target| target.tasks.contains_key(name))
-                {
-                    return Ok((self.project.default_environment(), task));
+            let default_env = self.project.default_environment();
+            // If the default environment has the task
+            if let Ok(default_env_task) = default_env.task(name, self.platform) {
+                // Unique if no other environment has the task name but a different task
+                let is_unique_task = self.project.environments().iter().all(|env| {
+                    match env.task(name, self.platform) {
+                        Ok(task) => task == default_env_task,
+                        Err(_) => true,
+                    }
+                });
+
+                if is_unique_task {
+                    return Ok((default_env, default_env_task));
```

Comment:
I think it makes more sense to turn this expression around: NOT any other task. Instead of all environments all refer to this task. (faster to find the negative)

Reply:
Im not entirely happy that we have to compare `Task`s now because this now still fails:

```
[tasks]
bla = "echo foo"

[feature.other.tasks]
bla = "echo foo"

[environments]
other = ["other"]
```

Which is arguably indeed still ambiguous. I think you should look at the unique features of an environment and see if it contains a task with the same name.

Reply:
Now I'm lost, I though that is what you described. Could you give an toml/shell example of `If the task is in de default environment and there is not other task with the same name (in any environment), run the task in the default environment.`

I think I'm parsing it wrong

Reply:
```toml
[feature.other.tasks]
bla = "echo foo"

[environments]
default = ["other"]
other = ["other"]
```

Reply:
Assuming you want that to spawn the selector, that would go against pavels example right?

Reply:
No I think its the same? Let me explain over video!

Reply:
I think it now does what we all expect :sweat_smile: